### PR TITLE
Search job async polling and cancellation

### PR DESCRIPTION
**Pull request by luk-kaminski**
## Description
Search job async polling and cancellation.
/nocl

## Usage in FE
1. You start execution like you did before: `api/views/search/65d4c08fd2005683f3bd4d10/execute`
2. You pick `id` and `executing_node` from the response.
3. In order to poll for a status of a job, you call a new GET endpoint: `api/views/searchjobs/[executing_node]/[id]/status`
4. In order to cancel a job, you call a new DELETE endpoint: `api/views/searchjobs/[executing_node]/[id]/cancel`

## Motivation and Context
Before this change, async `execute` endpoint was not async at all - it was working in synchronous mode.
It meant that FE had to wait for the completion of execution and had no option to cancel the request, i.e. when refreshing the page.
With this change `execute` endpoint is asynchronous, new `status` and `cancel` endpoints are introduced.
Execution returns response quickly, FE can poll the status with proper endpoint and cancel the execution with proper endpoint.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

/jenkins-pr-deps https://github.com/Graylog2/graylog-plugin-enterprise/pull/6769
/nocl

### CODE REVIEW

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackend.java [hunk 1]
```diff
@@ -17,6 +17,7 @@
 package org.graylog.storage.elasticsearch7.views;
 
 import com.google.common.collect.Maps;
+import com.unboundid.util.NotNull;
```
Comment (janheise) [inline graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackend.java:20]: Is this the correct import?

File: graylog-storage-opensearch2/src/main/java/org/graylog/storage/opensearch2/views/OpenSearchBackend.java [hunk 2]
```diff
@@ -17,6 +17,7 @@
 package org.graylog.storage.opensearch2.views;
 
 import com.google.common.collect.Maps;
+import com.unboundid.util.NotNull;
```
Comment (janheise) [inline graylog-storage-opensearch2/src/main/java/org/graylog/storage/opensearch2/views/OpenSearchBackend.java:20]: same here, is this the correct import?

File: graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackend.java (hunk 1 above)
Comment (luk-kaminski) [reply graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackend.java]: Good catch, thx!

**Comment by todvora** [comment]
The search page now triggers huge amount of status request. Is there a reason why we keep repeating those, even if the multisearch is already done and finished?

image
Peek 2024-03-20 13-05

**Review by todvora — APPROVED** [review]
LGTM, awesome work!

**Review by dennisoelkers — COMMENTED** [review]
Not finished with review, just adding some initial inline comments.

**Comment by luk-kaminski** [comment]
> @luk-kaminski When I merge the latest `master`, fix the conflicts and start the server I get the following error:
> 
> ```
> /graylog-project-repos/graylog-plugin-enterprise/enterprise/src/main/java/org/graylog/plugins/report/rest/WidgetExecutor.java:87:51
> java: cannot find symbol
>   symbol:   method execute(java.lang.String,org.graylog.plugins.views.search.permissions.SearchUser,org.graylog.plugins.views.search.rest.ExecutionState)
>   location: variable searchExecutor of type org.graylog.plugins.views.search.engine.SearchExecutor
> ```
> 
> Can you check if you get the same one with the latest changes, please?

@linuspahl - I believe it is because of the fact that related enterprise PR had not merged latest master.
So I have solved conflicts here, merged master there: https://github.com/Graylog2/graylog-plugin-enterprise/pull/6769
... and it works for me.

Can you give it a try?

**Review by linuspahl — APPROVED** [review]
It seems to work well and do not break the existing functionality.
I observed that we sometimes we do not call the `cancel` endpoint, while the UI is still showing a loading spinner. But looking at the `status` request response it only seems to occur in cases where the execution actually already finished.

**Review by janheise — APPROVED** [review]
I have two small things I noticed while testing:
`SearchJonsStatusResource` -> cancel is not visible because of a missing Api Annotation
Also, both methods in the Resource don't check if parameters are non-null, which looks like it's hanging in Swagger

Both could be discussed (maybe they are intentional) and/or could be added in a follow up PR I think.

**Pull request merged**